### PR TITLE
Update text to clarify glob behavior.

### PR DESCRIPTION
**Pull request by tetron**
Permit glob to refer to absolute paths within the output directory.

Update text to specify behavior when there are symlinks.

Refs https://github.com/common-workflow-language/common-workflow-language/issues/207

- [x] conformance test using `ln -s` symlink against file in working dir; symlink to input File, not staged. Symlink to input File that is staged. All of the above, but the symlink name is different.

Discussion on CWL chat 2018.10.16

Re: symlinks: 

- There are platforms that shutdown container before globbing. 
- When the output files are on cloud storage there is no concept of a symlink, and can not be followed.

Michael's statement, which we agree to, is that user expectation is that symlinks will be followed, so the specification will mandate that symlinks are resolved as expected.

- [x] @kaushik-work will write some tests involving symlinks
- [x] @tetron will revise wording of specification

### CODE REVIEW

File: CommandLineTool.yml [hunk 1]
```diff
@@ -238,6 +239,32 @@ $graph:
...
+        path pattern (it does begin with a slash '/') then it must
+        refer to a path within the output directory.  It is an error
+        if any glob, either relative path containing '..' or an
+        absolute path, resolves to a path outside the output
+        directory.
+
+        As a special case, a glob may refer to a symlink which points
+        to an input file which is located outside the output
+        directory.  However, in that case the glob pattern itself
+        resolves to a path in the output directory.  It is an error if
```
Comment (mr-c) [inline CommandLineTool.yml]: I find this paragraph confusing

File: CommandLineTool.yml [hunk 2]
```diff
@@ -238,6 +239,31 @@ $graph:
...
+        A glob may match a file within the output directory which is
+        actually a symlink to another file.  It is an error if a
+        symlink in the output directory refers to any file or
+        directory outside the output directory, except for the case of
+        files staged by InitialWorkDirRequirement and only if staging
+        was implementated using symlinks to the original input files.
+
+        A glob which refers to a symlink which is relative within the
+        output directory should produce a File object which has the
+        `location` of the glob path (the symlink is not dereferenced).
```
Comment (mr-c) [inline CommandLineTool.yml]: Can you share the motivation for this?

Comment (tetron) [reply CommandLineTool.yml]: The principal of least surprise.  Also, if symlinks are not supported by the storage system, the symlink might be converted to a regular file.

Comment (mr-c) [reply CommandLineTool.yml]: Symlinks are part of POSIX ...

Comment (tetron) [reply CommandLineTool.yml]: Yes, but several implementations upload the output directory to cloud storage and do the globbing separately.

Comment (mr-c) [reply CommandLineTool.yml]: I can't think of any circumstance that a researcher want to get a symlink "file" and not the actual contents. The choice to use an object store is on the implementer; they can traverse the symlink for the user, if needed.

Comment (mr-c) [reply CommandLineTool.yml]: in summary: remove 'location' and `symlink is not dereferenced'; the key is that the basename matches the name of the symlink

Comment (ghost) [reply CommandLineTool.yml]: Looking at 

```
If the value of the glob is an absolute path pattern (it does begin with a slash '/') then it must refer to a path within the output directory.
```

Wouldn't it be simpler to just not accept absolute paths. Looks like we are - correctly - restricting access to the working (output) directory. This does not let people collect outputs from arbitrary directories in the container, which is ok - but there is a case for it. Right now for some tools I tend to have a post processing step involve moving files from where they are produced to the output directory tree.

In summary :) I suggest that we either let people glob from arbitrary locations, or only have glob expressions that are relative to the output directory.

File: CommandLineTool.yml [hunk 3]
```diff
@@ -238,6 +239,31 @@ $graph:
         return a string or an array of strings, which will then be evaluated as
         one or more glob patterns.  Must only match and return files which
         actually exist.
+
+        If the value of glob is a relative path pattern (does not
+        begin with a slash '/') then it is resolved relative to the
+        output directory.  If the value of the glob is an absolute
```
Comment (ghost) [inline CommandLineTool.yml:252]: I suggest that we either let people glob from arbitrary locations, or only have glob expressions that are relative to the output directory. To simplify things, perhaps we should only allow relative globs (i.e. only products under the output directory are capturable)

File: CommandLineTool.yml (hunk 2 above)
Comment (ghost) [reply CommandLineTool.yml]: @tetron was this change put in (do you agree with it). I agree with @mr-c on this.

**Comment by ghost** [comment]
- pushed to top -

File: CommandLineTool.yml (hunk 2 above)
Comment (tetron) [reply CommandLineTool.yml]: From video chat discussion:  `location` can be whatever the platform wants, the important thing is that `basename` is consistent with the *symlink* name, and not the name of the file the link points to.

File: conformance_tests.yaml [hunk 4]
```diff
@@ -1768,3 +1768,53 @@
...
+
+- job: tests/empty.json
+  should_fail: true
+  tool: tests/glob-path-error.cwl
+  doc: Test fail trying to glob outside output directory
+  tags: [ required ]
+
+- job: tests/empty.json
+  tool: tests/symlink.cwl
+  doc: symlink to file outside of working directory should be retrieved
```
Comment (mr-c) [inline conformance_tests.yaml]:
```suggestion
  doc: symlink to file outside of working directory should be retrieved
  should_fail: true
```

**Review by ghost — APPROVED** [review]
Yes, this test should fail. I'll fix the cwl and the doc

File: CommandLineTool.yml [hunk 5]
```diff
@@ -245,6 +246,37 @@ $graph:
...
+        are illegal.
+
+        A glob may match a path within the output directory which is
+        actually a symlink to another file.  In this case, the
+        expected behavior is for the resulting File object to take the
+        `basename` (and corresponding `nameroot` and `nameext`) of the
+        symlink.  The `location` of the File is implementation
+        dependent, but logically the File should have the same content
+        as the symlink target.  Platforms may stage output files to
+        cloud storage where that lack the concept of a symlink.  In
```
Comment (ghost) [inline CommandLineTool.yml]: @tetron typo I think

Comment (tetron) [reply CommandLineTool.yml]: you mean "where that"

File: CommandLineTool.yml [hunk 6]
```diff
@@ -245,6 +246,37 @@ $graph:
...
+        duplication) the File `location` may refer to the symlink
+        target.
+
+        It is an error if a symlink in the output directory (or any
+        symlink in a chain of links) refers to any file or directory
+        that is not under an input or output directory.
+
+        Implementations may shut down a container before globbing
+        output, so globs and expressions must not assume access to the
+        container filesystem except for declared input and output.
```
Comment (mr-c) [inline CommandLineTool.yml:278]: This is not actionable advice from a user perspective. I recommend that this paragraph be removed.

Basically it reads that you can only glob on files/directories that you've already declared as being outputs which are themselves declared via globs.

File: CommandLineTool.yml [hunk 7]
```diff
@@ -245,6 +246,37 @@ $graph:
...
         actually exist.
+
+        If the value of glob is a relative path pattern (does not
+        begin with a slash '/') then it is resolved relative to the
+        output directory.  If the value of the glob is an absolute
+        path pattern (it does begin with a slash '/') then it must
+        refer to a path within the output directory.  It is an error
+        if any glob resolves to a path outside the output directory.
+        Specifically this means globs with relative paths containing
+        '..' or absolute paths that refer outside the output directory
```
Comment (mr-c) [inline CommandLineTool.yml:257]: Relative symlinks can have `..` in the middle and still not refer to paths outside of the input or output directories. This should instead say `relative paths starting with '..' `
